File: lsde-ts/src/handler-registry.ts
// LSDE Dialog Engine — Handler registration + Tier 1/Tier 2 resolution

import type {
	BlueprintBlock, BlockHandler, BaseBlockContext,
	DialogHandler, ChoiceHandler, ConditionHandler, ActionHandler,
	SceneLifecycleHandler,
	ValidateNextBlockHandler, InvalidateBlockHandler, BeforeBlockHandler,
} from './types.js';
import { BlockType } from './types.js';

// ─── Tier 1 — Global Registry ────────────────────────────────────────────────

/** Stores global (Tier 1) handlers. Last-write-wins per slot. */
export class HandlerRegistry {

	dialogHandler: DialogHandler | null = null;
	choiceHandler: ChoiceHandler | null = null;
	conditionHandler: ConditionHandler | null = null;
	actionHandler: ActionHandler | null = null;

	sceneEnterHandler: SceneLifecycleHandler | null = null;
	sceneExitHandler: SceneLifecycleHandler | null = null;

	validateNextBlockHandler: ValidateNextBlockHandler | null = null;
	invalidateBlockHandler: InvalidateBlockHandler | null = null;
	beforeBlockHandler: BeforeBlockHandler | null = null;

	getTypeHandler( type: BlockType ): BlockHandler<BlueprintBlock, BaseBlockContext> | null {
		switch ( type ) {
			case BlockType.Dialog: return this.dialogHandler as BlockHandler<BlueprintBlock, BaseBlockContext> | null;
			case BlockType.Choice: return this.choiceHandler as BlockHandler<BlueprintBlock, BaseBlockContext> | null;
			case BlockType.Condition: return this.conditionHandler as BlockHandler<BlueprintBlock, BaseBlockContext> | null;
			case BlockType.Action: return this.actionHandler as BlockHandler<BlueprintBlock, BaseBlockContext> | null;
			// A note is never dispatched, and a type this engine does not know is not either.
			default: return null;
		}
	}
}

// ─── Tier 2 — Per-Scene Registry ─────────────────────────────────────────────

/** Stores per-scene (Tier 2) handlers. */
export class SceneHandlerRegistry {

	private readonly blockHandlers = new Map<string, BlockHandler<BlueprintBlock, BaseBlockContext>>();

	dialogHandler: DialogHandler | null = null;
	choiceHandler: ChoiceHandler | null = null;
	conditionHandler: ConditionHandler | null = null;
	actionHandler: ActionHandler | null = null;

	enterHandler: SceneLifecycleHandler | null = null;
	exitHandler: SceneLifecycleHandler | null = null;

	setBlockHandler( blockId: string, handler: BlockHandler<BlueprintBlock, BaseBlockContext> ): void {
		this.blockHandlers.set( blockId, handler );
	}

	getBlockHandler( blockId: string ): BlockHandler<BlueprintBlock, BaseBlockContext> | null {
		return this.blockHandlers.get( blockId ) ?? null;
	}

	getTypeHandler( type: BlockType ): BlockHandler<BlueprintBlock, BaseBlockContext> | null {
		switch ( type ) {
			case BlockType.Dialog: return this.dialogHandler as BlockHandler<BlueprintBlock, BaseBlockContext> | null;
			case BlockType.Choice: return this.choiceHandler as BlockHandler<BlueprintBlock, BaseBlockContext> | null;
			case BlockType.Condition: return this.conditionHandler as BlockHandler<BlueprintBlock, BaseBlockContext> | null;
			case BlockType.Action: return this.actionHandler as BlockHandler<BlueprintBlock, BaseBlockContext> | null;
			// A note is never dispatched, and a type this engine does not know is not either.
			default: return null;
		}
	}
}

// ─── Resolution ──────────────────────────────────────────────────────────────

export interface ResolvedHandlers {
	sceneHandler: BlockHandler<BlueprintBlock, BaseBlockContext> | null;
	globalHandler: BlockHandler<BlueprintBlock, BaseBlockContext> | null;
}

/**
 * Resolve which handlers to call for a given block.
 * Priority: onBlock(id) > scene.onType > engine.onType
 */
export function resolveHandler(
	blockType: BlockType,
	blockId: string,
	sceneRegistry: SceneHandlerRegistry | null,
	globalRegistry: HandlerRegistry,
): ResolvedHandlers {
	const globalHandler = globalRegistry.getTypeHandler( blockType );

	if ( !sceneRegistry ) {
		return { sceneHandler: null, globalHandler };
	}

	// Most specific: onBlock(id)
	const blockOverride = sceneRegistry.getBlockHandler( blockId );
	if ( blockOverride ) {
		return { sceneHandler: blockOverride, globalHandler };
	}

	// Scene type override
	const sceneTypeHandler = sceneRegistry.getTypeHandler( blockType );
	if ( sceneTypeHandler ) {
		return { sceneHandler: sceneTypeHandler, globalHandler };
	}

	return { sceneHandler: null, globalHandler };
}

File: lsde-csharp/Runtime/HandlerRegistry.cs
// LSDE Dialog Engine — Handler registration + Tier 1/Tier 2 resolution (C# port of handler-registry.ts)

using System;
using System.Collections.Generic;

namespace LsdeDialogEngine
{
    // Internal non-generic handler delegate used by the traversal loop
    internal delegate Action? InternalBlockHandler(ISceneHandle scene, BlueprintBlock block, IBaseBlockContext context, Action next);

    /// <summary>Stores global (Tier 1) handlers. Last-write-wins per slot.</summary>
    internal class HandlerRegistry
    {
        internal BlockHandler<BlueprintBlock, IDialogContext>? DialogHandler;
        internal BlockHandler<BlueprintBlock, IChoiceContext>? ChoiceHandler;
        internal BlockHandler<BlueprintBlock, IConditionContext>? ConditionHandler;
        internal BlockHandler<BlueprintBlock, IActionContext>? ActionHandler;

        internal SceneLifecycleHandler? SceneEnterHandler;
        internal SceneLifecycleHandler? SceneExitHandler;

        internal ValidateNextBlockHandler? ValidateNextBlockHandler;
        internal InvalidateBlockHandler? InvalidateBlockHandler;
        internal BeforeBlockHandler? BeforeBlockHandler;

        internal InternalBlockHandler? GetTypeHandler(string type)
        {
            switch (type)
            {
                case BlockType.Dialog:
                    return DialogHandler != null ? WrapHandler(DialogHandler) : null;
                case BlockType.Choice:
                    return ChoiceHandler != null ? WrapHandler(ChoiceHandler) : null;
                case BlockType.Condition:
                    return ConditionHandler != null ? WrapHandler(ConditionHandler) : null;
                case BlockType.Action:
                    return ActionHandler != null ? WrapHandler(ActionHandler) : null;
                default:
                    // A note is never dispatched, and a type this engine does not know is not either.
                    return null;
            }
        }

        private static InternalBlockHandler WrapHandler<TBlock, TContext>(BlockHandler<TBlock, TContext> handler)
            where TBlock : BlueprintBlock
            where TContext : IBaseBlockContext
        {
            return (scene, block, context, next) =>
                handler(new BlockHandlerArgs<TBlock, TContext>(scene, (TBlock)block, (TContext)context, next));
        }
    }

    /// <summary>Stores per-scene (Tier 2) handlers.</summary>
    internal class SceneHandlerRegistry
    {
        private readonly Dictionary<string, InternalBlockHandler> _blockHandlers = new Dictionary<string, InternalBlockHandler>();

        internal BlockHandler<BlueprintBlock, IDialogContext>? DialogHandler;
        internal BlockHandler<BlueprintBlock, IChoiceContext>? ChoiceHandler;
        internal BlockHandler<BlueprintBlock, IConditionContext>? ConditionHandler;
        internal BlockHandler<BlueprintBlock, IActionContext>? ActionHandler;

        internal SceneLifecycleHandler? EnterHandler;
        internal SceneLifecycleHandler? ExitHandler;

        internal void SetBlockHandler(string blockId, InternalBlockHandler handler)
        {
            _blockHandlers[blockId] = handler;
        }

        internal InternalBlockHandler? GetBlockHandler(string blockId)
        {
            return _blockHandlers.TryGetValue(blockId, out var handler) ? handler : null;
        }

        internal InternalBlockHandler? GetTypeHandler(string type)
        {
            switch (type)
            {
                case BlockType.Dialog:
                    return DialogHandler != null ? WrapHandler(DialogHandler) : null;
                case BlockType.Choice:
                    return ChoiceHandler != null ? WrapHandler(ChoiceHandler) : null;
                case BlockType.Condition:
                    return ConditionHandler != null ? WrapHandler(ConditionHandler) : null;
                case BlockType.Action:
                    return ActionHandler != null ? WrapHandler(ActionHandler) : null;
                default:
                    // A note is never dispatched, and a type this engine does not know is not either.
                    return null;
            }
        }

        private static InternalBlockHandler WrapHandler<TBlock, TContext>(BlockHandler<TBlock, TContext> handler)
            where TBlock : BlueprintBlock
            where TContext : IBaseBlockContext
        {
            return (scene, block, context, next) =>
                handler(new BlockHandlerArgs<TBlock, TContext>(scene, (TBlock)block, (TContext)context, next));
        }
    }

    /// <summary>Resolved handler pair (scene + global).</summary>
    internal class ResolvedHandlers
    {
        internal InternalBlockHandler? SceneHandler;
        internal InternalBlockHandler? GlobalHandler;
    }

    internal static class HandlerResolver
    {
        /// <summary>
        /// Resolve which handlers to call for a given block.
        /// Priority: OnBlock(blockId) > scene.OnType > engine.OnType
        /// </summary>
        internal static ResolvedHandlers ResolveHandler(
            string blockType,
            string blockId,
            SceneHandlerRegistry? sceneRegistry,
            HandlerRegistry globalRegistry)
        {
            var globalHandler = globalRegistry.GetTypeHandler(blockType);

            if (sceneRegistry == null)
            {
                return new ResolvedHandlers { SceneHandler = null, GlobalHandler = globalHandler };
            }

            // Most specific: OnBlock(blockId)
            var blockOverride = sceneRegistry.GetBlockHandler(blockId);
            if (blockOverride != null)
            {
                return new ResolvedHandlers { SceneHandler = blockOverride, GlobalHandler = globalHandler };
            }

            // Scene type override
            var sceneTypeHandler = sceneRegistry.GetTypeHandler(blockType);
            if (sceneTypeHandler != null)
            {
                return new ResolvedHandlers { SceneHandler = sceneTypeHandler, GlobalHandler = globalHandler };
            }

            return new ResolvedHandlers { SceneHandler = null, GlobalHandler = globalHandler };
        }
    }
}
